create tfc credential secret objects in ocp namespaces

junminahn opened the pull request:
- create Terraform Cloud credential secret objects in all OCP namespaces requiring buckets
- load credentials data from credentials.json file
- add bash script linter (spellcheck)
- add bash script tests (Bats)

wenzowski commented:
I see Terratest running on CircleCI. What `TF_VAR_credentials`, `TF_VAR_project_name`, and `TF_VAR_test` are being used on `bcgov` CircleCI?

wenzowski commented on README.md:
```diff
@@ -24,6 +24,14 @@ make install_asdf_tools
+1. Navigate to IAM & Admin > Service Accounts page.
+1. Select a Service Account or create a new Service Account.
+   - The Service Account requires the roles `Storage Admin`, `Create Service Accounts`, `Delete Service Accounts`, and `Service Account Key Admin`.
+1. Create a Service Account key as JSON and download it in the root directory named `credentials.json`
```
I see this was added to .gitignore, great!

wenzowski commented on bcgov/main.tf:
```diff
@@ -58,7 +58,7 @@ resource "kubernetes_secret" "secret_sa" {
   metadata {
     name      = "gcp-${google_storage_bucket.bucket[count.index].name}-service-account-key"
-    namespace = "${split(",", element(var.namespace_apps, count.index))[0]}"
+    namespace = split(",", element(var.namespace_apps, count.index))[0]
```
If we're having to split this anyway, why not have two arguments? Not necessarily asking for a change but the thought process that led to this implementation.

junminahn replied:
two separate arguments for
namespace_apps and namespaces or
namespaces and apps ?

wenzowski replied:
Perhaps I misunderstood this line?
> it requires one string argument includes namespace and app name separated by comma, `"namespace,app"` in `command` job attribute

I'm asking why not send `"namespace" "app"` as arguments to the job. Can we explain any design constraints that led us to go with `"namespace,app"`?

junminahn replied:
I see your point, one of the terraform variable, namespace_apps, is an array of string separate by comma. so I used it as it is for the arguments but we can definitely allow two separate arguments for namespace and app and concatenate those two strings inside the script to add `namespace,app` value into `namespace_apps` variable.

wenzowski commented on openshift/deploy/job/cas-shelf-tfe-add-app.yml:
```diff
@@ -18,7 +18,7 @@ spec:
               cpu: 500m
               memory: 512Mi
-          image: docker.pkg.github.com/bcgov/cas-shelf/cas-shelf-tfe-add-app:0.1.0
+          image: docker.pkg.github.com/bcgov/cas-shelf/cas-shelf-tfe-add-app:0.1
```
Have we already added a note here on our tagging convention? The removal of patchlevel to refer only to `major.minor` release might not otherwise make sense to a reader.

junminahn replied:
hm.. whenever pushing a new image actually it overrides upper-level tag images as well.
and my intention was here to just ignore patch level and use whatever pushed the latest onto a minor level.
I will add a comment regarding the tag convention though

wenzowski replied:
> hm.. whenever pushing a new image actually it overrides upper-level tag images as well.

that totally makes sense, and since we're handing this code over that might not be obvious to another developer without documentation

junminahn commented:
> I see Terratest running on CircleCI. What `TF_VAR_credentials`, `TF_VAR_project_name`, and `TF_VAR_test` are being used on `bcgov` CircleCI?

wenzowski commented on tfe-scripts/tf-update-variable-values.sh:
```diff
@@ -20,14 +20,61 @@ else
+
+  if [ ! -z "$var_key" ] && [ ! -z "$var_val" ]; then
+    if [ $var_key == "namespace_apps" ]; then
+      # create double quote escaped namespace_apps array string and collect namespaces
```
Let's make sure we lint this and all other bash scripts with bats? Quoting is hard.
Example: https://github.com/bcgov/cas-pipeline/blob/master/.circleci/config.yml#L16

junminahn replied:
I will see how I could add Bats tests here

wenzowski replied:
sorry, silly me. Bats is our unit testing harness and Shellcheck is our bash linter. https://github.com/bcgov/cas-pipeline/blob/master/.circleci/config.yml#L11

Would be super ultra cool if we unit tested these scripts too ;)

junminahn commented:
`TF_VAR_test` is something I added when testing terraform variable and I just removed from CircleCI.

`TF_VAR_credentials` is credentials json string, usually Terraform takes a file path for GCP service account key but it is also possible to pass it as json string instead.

`TF_VAR_project_name` is GCP project_id, I changed `project_name` to `project_id` for the main scripts now but I used to call it project_name and all test scripts still follows the convention.

wenzowski commented on tfe-scripts/tf-update-variable-values.sh:
```diff
@@ -20,14 +20,61 @@ else
+      i=0
+      for value in $(echo "${var_val}" | jq -r '.[]'); do
+        namespaces+=( "$(cut -d',' -f1 <<<"$value")" )
+        if [ $i -ne 0 ]; then namespace_apps=$namespace_apps","; fi
```
```suggestion
        if [ $i -ne 0 ]; then namespace_apps="${namespace_apps},"; fi
```
possible quoting improvement (why I'm suggesting linter)

junminahn replied:
I will add Bats tests here as well!

wenzowski commented:
> so, `credentials` and `project_name` are being used for testing and necessary.

Cool, which service are they connected to? Let's make sure if it's running on BC Gov's CircleCI that it's pointing to instances that BC Gov owns and document which assets those are. If we need BC Gov to purchase assets for testing we'll need to surface that requirement.

matthieu-foucault commented on tfe-scripts/test/unit/add-item.bats:
```diff
@@ -0,0 +1,24 @@
+  list="$(echo "$data" | jq -r ".attributes.value")"
+  result="$(echo "$list" | jq ". + [\"$NEW_ITEM\"] | unique | @base64")"
+  echo "output = $result"
+  [ "$result" == "$EXPECTED" ]
```
This test needs to be changed to directly test your script instead of copying the jq query here. However, I don't think that this is something that needs to be done now. Please create a card to schedule this for a future sprint, as there are higher priority items that need to be taken care of.

junminahn replied:
I agree with that, I was thinking of creating a function to process that logic to be used for the scripts and tests, but as I removed much complexity by using jq instead of loop and escaped quotes, I didn't do it somehow.

junminahn commented on .circleci/config.yml:
```diff
@@ -49,6 +58,17 @@ jobs:
+      - run:
+          name: "Pull Submodules"
+          command: |
+            git submodule init
+            git submodule update --remote
```
I believe I copied this one from cas-postgres
https://github.com/bcgov/cas-postgres/blob/8457c421ea1e0f4e498e9e30eaca3e2a9b4807cc/.circleci/config.yml#L21

I will just take your suggestion here

junminahn commented on tfe-scripts/tf-add-app-run.sh:
```diff
@@ -1,81 +1,78 @@
-  RESULT=($(curl \
-    --header "Authorization: Bearer $TFC_TOKEN" \
-    "https://app.terraform.io/api/v2/runs/$RUN_ID" | jq -r '. | @base64'))
+# shellcheck disable=SC2016
```
I will add them

junminahn commented on tfe-scripts/tf-add-app-run.sh:
```diff
@@ -1,81 +1,78 @@
 
-if [ -z "$1" ]; then
-  echo "Usage: $0 <item>"
+if [ -z "$1" ] || [ -z "$2" ]; then
```
awesome! I will update others as well

junminahn commented on tfe-scripts/tf-add-app-run.sh:
```diff
@@ -1,81 +1,78 @@
-  echo "Usage: $0 <item>"
+if [ -z "$1" ] || [ -z "$2" ]; then
+  echo "Usage: $0 <namespace> <app>"
   exit 0
```
my initial convention actually came from Terraform API example to upload configurations.
https://www.terraform.io/docs/cloud/run/api.html#complete-script

but, I see your point as well!

junminahn commented on tfe-scripts/tf-add-app-run.sh:
```diff
@@ -1,81 +1,78 @@
+
+VAR_ID2="$(update_var "$TFC_WORKSPACE_ID" "$VAR_ID2" "$DATA2" | jq -r '.data.id')"
+
+echo "$VAR_ID2"
```
it is just for debugging for now

junminahn commented on tfe-scripts/tf-add-app-run.sh:
```diff
@@ -1,81 +1,78 @@
+NAMESPACE_APP="${NAMESPACE},${APP}"
+
+VAR1="namespace_apps"
+VAR2="kubernetes_namespaces"
```
I totally agree with the descriptive naming, but I might need to think about when to use uppercase variables and lowercase variables in the scripts.
I will change the variable names for sure!
